src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -207,66 +207,69 @@ private ProjectId GetActiveContextProjectId(IVsHierarchy hierarchy, IEnumerable<
                 var projectToHierarchyMap = _workspace._projectToHierarchyMap;
                 var solution = _workspace.CurrentSolution;
 
-                // We now must chase to the actual hierarchy that we know about. We'll do this as a loop as there may be multiple steps in order.
-                // intermediateHierarchy will be where we are so far, and we'll keep track of all of our intermediate steps (think a breadcrumb trail)
-                // in intermediateHierarchies.
-                var intermediateHierarchy = hierarchy;
-                var intermediateHierarchies = new HashSet<IVsHierarchy>();
-
+                // We now must chase to the actual hierarchy that we know about. First, we'll chase through multiple shared asset projects if
+                // we need to do so.
                 while (true)
                 {
-                    if (!intermediateHierarchies.Add(intermediateHierarchy))
+                    var contextHierarchy = hierarchy.GetActiveProjectContext();
+
+                    // The check for if contextHierarchy == hierarchy is working around downstream impacts of https://devdiv.visualstudio.com/DevDiv/_git/CPS/pullrequest/158271
+                    // Since that bug means shared projects have themselves as their own owner, it sometimes results in us corrupting state where we end up
+                    // having the context of shared project be itself, it seems.
+                    if (contextHierarchy == null || contextHierarchy == hierarchy)
                     {
-                        // We ended up somewhere we already were -- either we have a loop or we weren't able to make further progress. In this case,
-                        // just bail.
                         break;
                     }
 
-                    // Have we already arrived at a hierarchy we know about?
-                    var matchingProjectId = projectToHierarchyMap.FirstOrDefault(d => projectIds.Contains(d.Key) &&
-                                                                                      d.Value == intermediateHierarchy).Key;
+                    EnsureSubscribedToHierarchyPropertyChanges(hierarchy);
+                    hierarchy = contextHierarchy;
+                }
 
-                    if (matchingProjectId != null)
-                    {
-                        return matchingProjectId;
-                    }
+                // We may have multiple projects with the same hierarchy, but we can use __VSHPROPID8.VSHPROPID_ActiveIntellisenseProjectContext to distinguish
+                if (ErrorHandler.Succeeded(hierarchy.GetProperty(VSConstants.VSITEMID_ROOT, (int)__VSHPROPID8.VSHPROPID_ActiveIntellisenseProjectContext, out object contextProjectNameObject)))
+                {
+                    EnsureSubscribedToHierarchyPropertyChanges(hierarchy);
```

Comment:
❔ Why is this only inside the `ErrorHandler.Succeeded` guard? The intermediate hierarchies seem to get monitored unconditionally.

Reply:
(Warning: GitHub is flagging your comment as outdated; it's possible you're commenting on an older commit that I might have already fixed.)

The intermediate ones are only getting watched if they actually gave a pointer to another IVsHierarchy. Basically, you can think of this code as taking an IVsHierarchy and following pointers to other IVsHierarchies or projects. We're only sticking watchers on IVsHierarchies that gave us a redirection, so we'll know if they redirect elsewhere.